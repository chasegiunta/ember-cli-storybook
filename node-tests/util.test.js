const fs = require('fs');
const path = require('path');
const {
  generatePreviewHead,
  objectToHTMLAttributes,
  findEnvironment,
  overrideEnvironment,
  extendEnvironment,
  parse
} = require('../lib/util');

describe('util', () => {
  describe('objectToHTMLAttributes', () => {
    it('should be able to turn object into html attributes', () => {
      expect.assertions(1);
  
      expect(objectToHTMLAttributes({
  src: 'http://foo.com/?hi',
  rel: 'stylesheet' })).
toMatchInlineSnapshot(`"src=\\"http://foo.com/?hi\\" rel=\\"stylesheet\\""`);
    });
  });

  describe('@parse', () => {
    it('should be able to parse built html file', () => {
      expect.assertions(1);
  
      const fileContent = fs.readFileSync(path.resolve(__dirname, 'fixtures', 'build.html'), 'utf8');
  
      expect(parse(fileContent)).toMatchInlineSnapshot(`
Object {
  "link": Array [
    Object {
      "href": "./assets/vendor.css",
      "rel": "stylesheet",
    },
    Object {
      "href": "./assets/storybook-ember-3-1.css",
      "rel": "stylesheet",
    },
    Object {
      "href": "./assets/test-support.css",
      "rel": "stylesheet",
    },
  ],
  "meta": Array [
    Object {
      "content": "%7B%22modulePrefix%22%3A%22storybook-ember-3-1%22%2C%22environment%22%3A%22test%22%2C%22rootURL%22%3A%22/%22%2C%22locationType%22%3A%22none%22%2C%22EmberENV%22%3A%7B%22FEATURES%22%3A%7B%7D%2C%22EXTEND_PROTOTYPES%22%3A%7B%22Date%22%3Afalse%7D%7D%2C%22APP%22%3A%7B%22LOG_ACTIVE_GENERATION%22%3Afalse%2C%22LOG_VIEW_LOOKUPS%22%3Afalse%2C%22rootElement%22%3A%22%23ember-testing%22%2C%22autoboot%22%3Afalse%2C%22name%22%3A%22storybook-ember-3-1%22%2C%22version%22%3A%220.0.0+eebe77e5%22%7D%2C%22exportApplicationGlobal%22%3Atrue%7D",
      "name": "storybook-ember-3-1/config/environment",
    },
  ],
  "script": Array [
    Object {
      "src": "./testem.js",
    },
    Object {
      "src": "./assets/vendor.js",
    },
    Object {
      "src": "./assets/test-support.js",
    },
    Object {
      "src": "./assets/storybook-ember-3-1.js",
    },
    Object {
      "src": "./assets/tests.js",
    },
    Object {
      "src": "./assets/component.js",
      "type": "module",
    },
  ],
}
`);
    });
  
  
    it('should strip rootURL found in config href and src attributes', () => {
      expect.assertions(1);
  
      const fileContent = fs.readFileSync(path.resolve(__dirname, 'fixtures', 'root-url.html'), 'utf8');
  
      expect(parse(fileContent)).toMatchInlineSnapshot(`
Object {
  "link": Array [
    Object {
      "href": "./assets/vendor.css",
      "rel": "stylesheet",
    },
    Object {
      "href": "./assets/vault.css",
      "rel": "stylesheet",
    },
    Object {
      "href": "./favicon.png",
      "rel": "icon",
    },
    Object {
      "href": "http://example.com/external/style.css",
      "rel": "stylesheet",
    },
  ],
  "meta": Array [
    Object {
      "content": "%7B%22modulePrefix%22%3A%22vault%22%2C%22environment%22%3A%22development%22%2C%22rootURL%22%3A%22/ui/%22%2C%22locationType%22%3A%22auto%22%2C%22EmberENV%22%3A%7B%22FEATURES%22%3A%7B%7D%2C%22EXTEND_PROTOTYPES%22%3A%7B%22Date%22%3Afalse%7D%7D%2C%22APP%22%3A%7B%22POLLING_URLS%22%3A%5B%22sys/health%22%2C%22sys/replication/status%22%2C%22sys/seal-status%22%5D%2C%22NAMESPACE_ROOT_URLS%22%3A%5B%22sys/health%22%2C%22sys/seal-status%22%2C%22sys/license/features%22%5D%2C%22DEFAULT_PAGE_SIZE%22%3A15%2C%22LOG_TRANSITIONS%22%3Atrue%7D%2C%22flashMessageDefaults%22%3A%7B%22timeout%22%3A7000%2C%22sticky%22%3Afalse%2C%22preventDuplicates%22%3Atrue%7D%2C%22contentSecurityPolicyHeader%22%3A%22Content-Security-Policy%22%2C%22contentSecurityPolicyMeta%22%3Atrue%2C%22contentSecurityPolicy%22%3A%7B%22connect-src%22%3A%5B%22%27self%27%22%5D%2C%22img-src%22%3A%5B%22%27self%27%22%2C%22data%3A%22%5D%2C%22form-action%22%3A%5B%22%27none%27%22%5D%2C%22script-src%22%3A%5B%22%27self%27%22%5D%2C%22style-src%22%3A%5B%22%27unsafe-inline%27%22%2C%22%27self%27%22%5D%2C%22default-src%22%3A%5B%22%27none%27%22%5D%2C%22font-src%22%3A%5B%22%27self%27%22%5D%2C%22media-src%22%3A%5B%22%27self%27%22%5D%7D%2C%22emberData%22%3A%7B%22enableRecordDataRFCBuild%22%3Afalse%7D%2C%22exportApplicationGlobal%22%3Atrue%7D",
      "name": "vault/config/environment",
    },
  ],
  "script": Array [
    Object {
      "src": "./assets/vendor.js",
    },
    Object {
      "src": "./assets/vault.js",
    },
    Object {
      "src": "http://example.com/external/script.js",
    },
    Object {
      "src": "./ember-cli-live-reload.js",
      "type": "text/javascript",
    },
  ],
}
`);
    });
  
    it('should be able to parse built html file and strip out test related files', () => {
      expect.assertions(1);
  
      const fileContent = fs.readFileSync(path.resolve(__dirname, 'fixtures', 'build.html'), 'utf8');
  
      expect(parse(fileContent, true)).toMatchInlineSnapshot(`
Object {
  "link": Array [
    Object {
      "href": "./assets/vendor.css",
      "rel": "stylesheet",
    },
    Object {
      "href": "./assets/storybook-ember-3-1.css",
      "rel": "stylesheet",
    },
  ],
  "meta": Array [
    Object {
      "content": "%7B%22modulePrefix%22%3A%22storybook-ember-3-1%22%2C%22environment%22%3A%22test%22%2C%22rootURL%22%3A%22/%22%2C%22locationType%22%3A%22none%22%2C%22EmberENV%22%3A%7B%22FEATURES%22%3A%7B%7D%2C%22EXTEND_PROTOTYPES%22%3A%7B%22Date%22%3Afalse%7D%7D%2C%22APP%22%3A%7B%22LOG_ACTIVE_GENERATION%22%3Afalse%2C%22LOG_VIEW_LOOKUPS%22%3Afalse%2C%22rootElement%22%3A%22%23ember-testing%22%2C%22autoboot%22%3Afalse%2C%22name%22%3A%22storybook-ember-3-1%22%2C%22version%22%3A%220.0.0+eebe77e5%22%7D%2C%22exportApplicationGlobal%22%3Atrue%7D",
      "name": "storybook-ember-3-1/config/environment",
    },
  ],
  "script": Array [
    Object {
      "src": "./testem.js",
    },
    Object {
      "src": "./assets/vendor.js",
    },
    Object {
      "src": "./assets/storybook-ember-3-1.js",
    },
    Object {
      "src": "./assets/component.js",
      "type": "module",
    },
  ],
}
`);
    });

    it('should be able to parse metas from a built html file from an app that uses engines', () => {
      expect.assertions(1);

      const fileContent = fs.readFileSync(path.resolve(__dirname, 'fixtures', 'engines.html'), 'utf8');
      const metas = parse(fileContent, true).meta;

      expect(metas).toMatchSnapshot();
    });
  });

  describe('@generatePreviewHead', () => {
    it('should work with file created with `ember build`', () => {
      expect.assertions(1);

      const fileContent = fs.readFileSync(path.resolve(__dirname, 'fixtures', 'build.html'), 'utf8');

      expect(generatePreviewHead(parse(fileContent))).toMatchSnapshot()
    })

    it('should work with file created with `ember build` in production env', () => {
      expect.assertions(1);

      const fileContent = fs.readFileSync(path.resolve(__dirname, 'fixtures', 'build-production.html'), 'utf8');

      expect(generatePreviewHead(parse(fileContent))).toMatchSnapshot()
    })

    it('should work with file created with `ember serve` (should append livereload pointing at serve instance)', () => {
      expect.assertions(1);

      const fileContent = fs.readFileSync(path.resolve(__dirname, 'fixtures', 'serve.html'), 'utf8');

      expect(generatePreviewHead(parse(fileContent))).toMatchSnapshot();
    });
  });

  describe('@findEnvironment', () => {
    it('should return the meta node with name config/environment (which has the app env encoded as its content)', () => {
      expect.assertions(1);

      const fileContent = fs.readFileSync(path.resolve(__dirname, 'fixtures', 'serve.html'), 'utf8');
      const config = parse(fileContent, true);
      const env = findEnvironment(config);

      expect(env.name).toBe('storybook-ember-3-1/config/environment');
    });

    it('should be robust against a file with no environment node', () => {
      expect.assertions(1);

      const fileContent = fs.readFileSync(path.resolve(__dirname, 'fixtures', 'no-env.html'), 'utf8');
      const config = parse(fileContent, true);
      const env = findEnvironment(config);

      expect(env).toBeUndefined();
    });
  });

  describe('@extendEnvironment', () => {
    it('deeply merges the original env object with override objects', () => {
      const env = {
        one: 'fish',
        two: {
          fishes: [ 'red', 'blue' ]
        }
      };

      expect(
        extendEnvironment(
          env,
          { two: { fishes: [ 'yellow', 'purple' ], updated: true } },
          { hello: 'world' }
        )
      ).toEqual({
        one: 'fish',
        two: {
          fishes: [ 'yellow', 'purple' ],
          updated: true,
        },
        hello: 'world',
      });
    });

    it('should be robust against undefined input', () => {
      const env = {
        one: 'fish',
        two: {
          fishes: [ 'red', 'blue' ]
        }
      };

      expect(extendEnvironment(env, undefined, undefined)).toEqual({
        one: 'fish',
        two: {
          fishes: [ 'red', 'blue' ]
        }
      });
    });
  });

  describe('@overrideEnvironment', () => {
    // Take a parsed fixture, get env node, override, assert strings as well as intermediates
    // Assert works fine with undefined argument
    it('returns a properly encoded meta value for an environment with overrides', () => {
      expect.assertions(1);

      const fileContent = fs.readFileSync(path.resolve(__dirname, 'fixtures', 'serve.html'), 'utf8');
      const config = parse(fileContent, true);
      const env = findEnvironment(config);
      const overridden = overrideEnvironment(env, {
        rootURL: '/foobar/',
        meta: 'data',
        APP: {
          name: 'new-name-from-test',
        },
      });

      const envObject = JSON.parse(decodeURIComponent(env.content));
      envObject.rootURL = '/foobar/';
      envObject.meta = 'data';
      envObject.APP.name = 'new-name-from-test';

      expect(overridden).toBe(encodeURIComponent(JSON.stringify(envObject)));
    });
  });
});

