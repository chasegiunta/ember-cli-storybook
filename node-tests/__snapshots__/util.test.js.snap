// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`util @generatePreviewHead should work with file created with \`ember build\` 1`] = `
"<!-- This file is auto-generated by ember-cli-storybook -->
<meta name=\\"storybook-ember-3-1/config/environment\\" content=\\"%7B%22modulePrefix%22%3A%22storybook-ember-3-1%22%2C%22environment%22%3A%22test%22%2C%22rootURL%22%3A%22/%22%2C%22locationType%22%3A%22none%22%2C%22EmberENV%22%3A%7B%22FEATURES%22%3A%7B%7D%2C%22EXTEND_PROTOTYPES%22%3A%7B%22Date%22%3Afalse%7D%7D%2C%22APP%22%3A%7B%22LOG_ACTIVE_GENERATION%22%3Afalse%2C%22LOG_VIEW_LOOKUPS%22%3Afalse%2C%22rootElement%22%3A%22%23ember-testing%22%2C%22autoboot%22%3Afalse%2C%22name%22%3A%22storybook-ember-3-1%22%2C%22version%22%3A%220.0.0+eebe77e5%22%7D%2C%22exportApplicationGlobal%22%3Atrue%7D\\" />
<link rel=\\"stylesheet\\" href=\\"./assets/vendor.css\\" />
<link rel=\\"stylesheet\\" href=\\"./assets/storybook-ember-3-1.css\\" />
<link rel=\\"stylesheet\\" href=\\"./assets/test-support.css\\" />
<script src=\\"./testem.js\\"></script>
<script src=\\"./assets/vendor.js\\"></script>
<script>window.Ember = require(\\"ember\\").default; runningTests = true; Ember.testing=true;</script>
<script src=\\"./assets/test-support.js\\"></script>
<script src=\\"./assets/storybook-ember-3-1.js\\"></script>
<script src=\\"./assets/tests.js\\"></script>
<script src=\\"./assets/component.js\\" type=\\"module\\"></script>"
`;

exports[`util @generatePreviewHead should work with file created with \`ember build\` in production env 1`] = `
"<!-- This file is auto-generated by ember-cli-storybook -->
<meta name=\\"before-storybook/config/environment\\" content=\\"%7B%22modulePrefix%22%3A%22before-storybook%22%2C%22environment%22%3A%22production%22%2C%22rootURL%22%3A%22%2F%22%2C%22locationType%22%3A%22history%22%2C%22EmberENV%22%3A%7B%22EXTEND_PROTOTYPES%22%3Afalse%2C%22FEATURES%22%3A%7B%7D%2C%22_APPLICATION_TEMPLATE_WRAPPER%22%3Afalse%2C%22_DEFAULT_ASYNC_OBSERVERS%22%3Atrue%2C%22_JQUERY_INTEGRATION%22%3Afalse%2C%22_TEMPLATE_ONLY_GLIMMER_COMPONENTS%22%3Atrue%7D%2C%22APP%22%3A%7B%22name%22%3A%22before-storybook%22%2C%22version%22%3A%220.0.0%22%7D%7D\\" />
<link rel=\\"stylesheet\\" href=\\"./assets/vendor-d41d8cd98f00b204e9800998ecf8427e.css\\" />
<link rel=\\"stylesheet\\" href=\\"./assets/before-storybook-d41d8cd98f00b204e9800998ecf8427e.css\\" />
<script src=\\"./assets/vendor-b6ca89ba21b4a11ce607688b017119f0.js\\"></script>
<script>window.Ember = require(\\"ember\\").default; runningTests = true; Ember.testing=true;</script>
<script src=\\"./assets/before-storybook-0c543b7d02d9385b2e4f5e2f3aab96f3.js\\"></script>"
`;

exports[`util @generatePreviewHead should work with file created with \`ember serve\` (should append livereload pointing at serve instance) 1`] = `
"<!-- This file is auto-generated by ember-cli-storybook -->
<meta name=\\"storybook-ember-3-1/config/environment\\" content=\\"%7B%22modulePrefix%22%3A%22storybook-ember-3-1%22%2C%22environment%22%3A%22test%22%2C%22rootURL%22%3A%22/%22%2C%22locationType%22%3A%22none%22%2C%22EmberENV%22%3A%7B%22FEATURES%22%3A%7B%7D%2C%22EXTEND_PROTOTYPES%22%3A%7B%22Date%22%3Afalse%7D%7D%2C%22APP%22%3A%7B%22LOG_ACTIVE_GENERATION%22%3Afalse%2C%22LOG_VIEW_LOOKUPS%22%3Afalse%2C%22rootElement%22%3A%22%23ember-testing%22%2C%22autoboot%22%3Afalse%2C%22name%22%3A%22storybook-ember-3-1%22%2C%22version%22%3A%220.0.0+eebe77e5%22%7D%2C%22exportApplicationGlobal%22%3Atrue%7D\\" />
<link rel=\\"stylesheet\\" href=\\"./assets/vendor.css\\" />
<link rel=\\"stylesheet\\" href=\\"./assets/storybook-ember-3-1.css\\" />
<link rel=\\"stylesheet\\" href=\\"./assets/test-support.css\\" />
<script src=\\"./testem.js\\"></script>
<script src=\\"./assets/vendor.js\\"></script>
<script>window.Ember = require(\\"ember\\").default; runningTests = true; Ember.testing=true;</script>
<script src=\\"./assets/test-support.js\\"></script>
<script src=\\"./assets/storybook-ember-3-1.js\\"></script>
<script src=\\"./assets/tests.js\\"></script>
<script>
            (function() {
              var srcUrl = null;
              var host = location.hostname || 'localhost';
              var defaultPort = location.protocol === 'https:' ? 443 : 80;
              var port = undefined;
              var path = '';
              var prefixURL = '';
              var src = srcUrl || prefixURL + '/_lr/livereload.js?port=' + port + '&host=' + host + path;
              var script = document.createElement('script');
              script.type = 'text/javascript';
              script.src = location.protocol + '//' + host + ':undefined' + src;
              document.getElementsByTagName('head')[0].appendChild(script);
            }());
          </script>
<script src=\\"./assets/component.js\\" type=\\"module\\"></script>"
`;

exports[`util @parse should be able to parse metas from a built html file from an app that uses engines 1`] = `
Array [
  Object {
    "content": "%7B%22modulePrefix%22%3A%22vault%22%2C%22environment%22%3A%22development%22%2C%22rootURL%22%3A%22/ui/%22%2C%22locationType%22%3A%22auto%22%2C%22EmberENV%22%3A%7B%22FEATURES%22%3A%7B%7D%2C%22EXTEND_PROTOTYPES%22%3A%7B%22Date%22%3Afalse%7D%7D%2C%22APP%22%3A%7B%22POLLING_URLS%22%3A%5B%22sys/health%22%2C%22sys/replication/status%22%2C%22sys/seal-status%22%5D%2C%22NAMESPACE_ROOT_URLS%22%3A%5B%22sys/health%22%2C%22sys/seal-status%22%2C%22sys/license/features%22%5D%2C%22DEFAULT_PAGE_SIZE%22%3A15%2C%22LOG_TRANSITIONS%22%3Atrue%7D%2C%22flashMessageDefaults%22%3A%7B%22timeout%22%3A7000%2C%22sticky%22%3Afalse%7D%2C%22contentSecurityPolicyHeader%22%3A%22Content-Security-Policy%22%2C%22contentSecurityPolicyMeta%22%3Atrue%2C%22contentSecurityPolicy%22%3A%7B%22connect-src%22%3A%5B%22%27self%27%22%5D%2C%22img-src%22%3A%5B%22%27self%27%22%2C%22data%3A%22%5D%2C%22form-action%22%3A%5B%22%27none%27%22%5D%2C%22script-src%22%3A%5B%22%27self%27%22%5D%2C%22style-src%22%3A%5B%22%27unsafe-inline%27%22%2C%22%27self%27%22%5D%2C%22default-src%22%3A%5B%22%27none%27%22%5D%2C%22font-src%22%3A%5B%22%27self%27%22%5D%2C%22media-src%22%3A%5B%22%27self%27%22%5D%7D%2C%22emberData%22%3A%7B%22enableRecordDataRFCBuild%22%3Afalse%7D%2C%22exportApplicationGlobal%22%3Atrue%7D",
    "name": "vault/config/environment",
  },
  Object {
    "content": "%7B%22modulePrefix%22%3A%22replication%22%2C%22environment%22%3A%22development%22%7D",
    "name": "replication/config/environment",
  },
  Object {
    "content": "%7B%22bundles%22%3A%7B%22replication%22%3A%7B%22assets%22%3A%5B%7B%22uri%22%3A%22/ui/engines-dist/replication/assets/engine-vendor.js%22%2C%22type%22%3A%22js%22%7D%2C%7B%22uri%22%3A%22/ui/engines-dist/replication/assets/engine.js%22%2C%22type%22%3A%22js%22%7D%5D%7D%7D%7D",
    "name": "vault/config/asset-manifest",
  },
]
`;
